### PR TITLE
SI-2197-vss-priv-history-endpoint

### PR DESCRIPTION
**Pull request by Allyson-English**
# Proposed Changes
New route: `v2/vehicle/{tokenId}/history`

### Notes
- recreates existing history endpoint but pointed at VSS schema
- requires same privileges
- Elasticsearch index is determined using path, ie:
```
DEVICE_DATA_INDEX_NAME:
    v1: device-status-dev*
    v2: vss-status-dev*
```

### CODE REVIEW

File: internal/controllers/device_data_controller.go [hunk 1]
```diff
@@ -215,10 +215,11 @@ func (d *DeviceDataController) getHistory(c *fiber.Ctx, userDevice *grpc.UserDev
 				Query: &types.Query{
 					Bool: &types.BoolQuery{
 						Filter: []types.Query{
-							{Term: map[string]types.TermQuery{"subject": {Value: userDevice.Id}}},
-							{Range: map[string]types.RangeQuery{"data.timestamp": types.DateRangeQuery{Gte: some.String(startDate), Lte: some.String(endDate)}}},
+							{Match: map[string]types.MatchQuery{"subject": {Query: userDevice.Id}}},
```
Comment (elffjs) [inline internal/controllers/device_data_controller.go:218]: Why did you have to do this?

File: internal/controllers/device_data_controller.go [hunk 2]
```diff
@@ -190,9 +190,9 @@ func (d *DeviceDataController) GetHistoricalRawPermissioned(c *fiber.Ctx) error
 	var filter types.SourceFilter
 
 	if slices.Contains(privileges, AllTimeLocation) {
-		filter.Includes = append(filter.Includes, "data.latitude", "data.longitude", "location", "data.cell", "cell")
+		filter.Includes = append(filter.Includes, "data.latitude", "data.longitude", "location", "data.cell", "cell", "data.vehicle.currentLocation")
 	} else {
-		filter.Excludes = append(filter.Excludes, "data.latitude", "data.longitude", "location", "data.cell", "cell")
+		filter.Excludes = append(filter.Excludes, "data.latitude", "data.longitude", "location", "data.cell", "cell", "data.vehicle.currentLocation")
```
Comment (elffjs) [inline internal/controllers/device_data_controller.go:195]: Doesn't this contain the timestamp?

File: cmd/device-data-api/api.go [hunk 3]
```diff
@@ -99,6 +99,15 @@ func startWebAPI(logger zerolog.Logger, settings *config.Settings, dbs func() *d
 	vToken.Get("/history", tk.OneOf(vehicleAddr, []int64{controllers.NonLocationData, controllers.AllTimeLocation}), cacheHandler, deviceDataController.GetHistoricalRawPermissioned)
 	vToken.Get("/status", tk.OneOf(vehicleAddr, []int64{controllers.NonLocationData, controllers.CurrentLocation, controllers.AllTimeLocation}), cacheHandler, deviceDataController.GetVehicleStatus)
 
+	// vss schema, v2
+	if settings.Environment != "prod" {
+		vehTokenV2 := app.Group("/v2/vehicle/:tokenID")
+		vssSettings := *settings
+		vssSettings.DeviceDataIndexName = "vss-status-dev*"
```
Comment (elffjs) [inline cmd/device-data-api/api.go:106]: This is a bit messy but I suppose it does work. What would we do on prod, though?

File: internal/controllers/device_data_controller.go (hunk 1 above)
Comment (Allyson-English) [reply internal/controllers/device_data_controller.go:218]: It seems like the term query searches for the exact value where the match query searches for any matches... my suspicion is that keywords are playing in a role in why the two queries turn up different results. 

Seems like the difference lies in how [Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-term-query.html#:~:text=To%20better%20search%20text%20fields,the%20exact%20term%20you%20provide) applies an analyzer for each respective query type. I found that `Match` returned the expected response for both indexes but `Term` didn't return anything for the vss index

Comment (elffjs) [reply internal/controllers/device_data_controller.go:218]: Then I think the problem is our mappings. Let me look.

File: charts/device-data-api/values-prod.yaml [hunk 4]
```diff
@@ -8,16 +8,18 @@ cronJob:
...
 env:
   ENVIRONMENT: prod
-  PORT: '8080'
+  PORT: "8080"
   LOG_LEVEL: info
   SERVICE_NAME: device-data-api
   JWT_KEY_SET_URL: https://auth.dimo.zone/keys
   DEPLOYMENT_BASE_URL: https://device-data-api.dimo.zone
-  DEVICE_DATA_INDEX_NAME: device-status-prod*
+  DEVICE_DATA_INDEX_NAME:
```
Comment (elffjs) [inline charts/device-data-api/values-prod.yaml:16]: I doubt this will work.

File: internal/controllers/device_data_controller.go [hunk 5]
```diff
@@ -293,12 +294,12 @@ func removeOdometerIfInvalid(body []byte) []byte {
 // @Router       /user/device-data/{userDeviceID}/distance-driven [get]
 func (d *DeviceDataController) GetDistanceDriven(c *fiber.Ctx) error {
 	userDeviceID := c.Params("userDeviceID")
-
-	odoStart, err := d.queryOdometer(c.Context(), sortorder.Asc, userDeviceID)
+	esIdx := d.Settings.DeviceDataIndexName[strings.Split(c.OriginalURL(), "/")[1]]
```
Comment (elffjs) [inline internal/controllers/device_data_controller.go:297]: Why are we changing this?

File: internal/controllers/device_data_controller.go [hunk 6]
```diff
@@ -104,7 +104,47 @@ func (d *DeviceDataController) GetHistoricalRaw(c *fiber.Ctx) error {
...
+// @Description  Get all historical data for a userDeviceID, within start and end range
+// @Tags         device-data
+// @Produce      json
+// @Success      200
+// @Param        userDeviceID  path   string  true   "user id"
+// @Param        startDate     query  string  false  "startDate eg 2022-01-02. if empty two weeks back"
+// @Param        endDate       query  string  false  "endDate eg 2022-03-01. if empty today"
+// @Security     BearerAuth
+// @Router       /user/device-data/{userDeviceID}/historical [get]
+func (d *DeviceDataController) GetHistoricalRawV2(c *fiber.Ctx) error {
```
Comment (elffjs) [inline internal/controllers/device_data_controller.go:120]: I don't think we should have this endpoint. Can we get rid of it?

Comment (elffjs) [reply internal/controllers/device_data_controller.go]: I did it.

Comment (elffjs) [reply internal/controllers/device_data_controller.go]: Did it.

**Review by elffjs — APPROVED** [review]

**Pull request merged**